src/server/main_service.cc

```diff
@@ -1229,8 +1226,9 @@ bool Service::InvokeCmd(const CommandId* cid, CmdArgList tail_args, ConnectionCo
   // TODO: we should probably discard more commands here,
   // not just the blocking ones
   const auto* conn = cntx->conn();
-  if (!(cid->opt_mask() & CO::BLOCKING) && conn != nullptr && etl.GetSlowLog().IsEnabled() &&
-      invoke_time_usec >= etl.log_slower_than_usec) {
+  if (!(cid->opt_mask() & CO::BLOCKING) && conn != nullptr &&
+      ServerState::tlocal()->GetSlowLog().IsEnabled() &&
```

Comment:
actually you are right, the compiler can not cache `ServerState::tlocal()` into a register since it does not know if `cid->Invoke` changes it.  Ok, please add a comment before line 1229 explaining that `ServerState::tlocal()` should not be cached into a local variable due to possible thread migration

Reply:
https://clang.llvm.org/docs/AttributeReference.html#optnone or `noinline` ?

Reply:
@dranikpg you are right. it's so confusing. so if we have 
```
thread_local int val;
void foo() {
 int i = val;
 bar();  // not inline, migrates
 int j = val;
}
```
`foo` can not know that `bar` does not modify val so it must reread it after the call, but it can cache the offset to `val` based on the base pointer of the original thread, so it will reread val but from the original location. 
So, @kostasrim , we can not rely on `tlocal()` calls and this is why we need `cntx->server_state`. 
now, I do not entirely understand why you tried to include server from facade, since `ConnectionContext` is located here:    https://github.com/dragonflydb/dragonfly/blob/main/src/server/conn_context.h#L166

Reply:
Besides InvokeCmd and places up the call stack like DispatchCommand / Squasher and Connection, we don't have any dangerous sections anymore.

We have 130(!) accesses to ServerState::tlocal(), most are in callbacks, which is fine (except the ones that call Migrate internally). Making this a field is kinda strange because it's (1) another solution and (2) if it's managed by the connection, it involves some trickery (because the connection doesn't have access to server state).

What if create a non-inlined `safe_tlocal()` function that we use in this few calls?

Reply:
>  I do not entirely understand why you tried to include server from facade, since ConnectionContext is located here: https://github.com/dragonflydb/dragonfly/blob/main/src/server/conn_context.h#L166

@romange Because the migration logic is in `facade/dragonfly_connection.cc` and we need to update the `ConnectionContext` there which is declared in `server/conn_context`. 

@dranikpg 100% agree, that's why I suggested `optnone` above. The question is if this actually works on not? I need to try this on godbolt

Reply:
noinline should be enough

Reply:
yes, we may use noinline attribute for `safe_tlocal` . Please note that defining a function in cc file is not enough because we use `-flto` optimization which may still inline it.

Reply:
> Please note that defining a function in cc file will not help because we use -flto optimization which may still inline it.

You mean this even if attribute `noinline` is used? (which implies definition in header via `static void __attribute__((noinline))`. 

I would be very surprised if this is the case

Reply:
I have not said this. I said  - just defining it in cc file.